geoproxy/main.go

```diff
@@ -272,87 +315,230 @@ func getSubmitProxyUrl(rawCoord []float64) string {
+	if id == uuid.Nil {
+		// Then we just have a submit request.
+		// We only have to send it to a single server.
+		// Just process it normally
+		if verbose {
+			log.Println("Making submit request")
+		}
+		url, _ := url.Parse(target[0])
+		proxy := httputil.NewSingleHostReverseProxy(url)
+		req.URL.Host = url.Host
+		req.URL.Scheme = url.Scheme
+		req.Host = url.Host
+		if verbose {
+			for header, values := range req.Header {
+		       		for _, value := range values {
+					log.Printf("head: %s, val: %s \n", header, value)
+			}
+			}
+	}
+		proxy.ServeHTTP(res, req)
+		return
+	}
 
-	// //Need to be able to handle OPTIONS, see https://flaviocopes.com/golang-enable-cors/ for details
-	// if (*req).Method == "OPTIONS" {
-	// 	return
-	// }
+	if verbose {
+		log.Println("Making view request")
+	}
+	// Read body to buffer
+	
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		log.Printf("Error reading body: %v", err)
+		return
+	}
+	
+	log.Printf("req url: %s \n", html.EscapeString(req.URL.Path))
+
+	// Edit request body to include id
+	buff := bytes.NewBuffer(body)
+	bodyStr := buff.String()
+	// This is where we want to insert the id param	
+	i := strings.LastIndex(bodyStr, "}")
+	// Convert to runes to split
+	runes := []rune(bodyStr)
+	// Left side string of \n}
+	leftStr := string(runes[0:i])
+	buff = bytes.NewBufferString(leftStr)
+	// Concatenate new id param and request body remainder
+	buff.WriteString(",\"id\":" + "\"" + id.String() + "\"")
+	buff.WriteString(string(runes[i:len(runes)]))
+	buffBytes := buff.Bytes()	
+	
+	setupTimer(id)
+	//fmt.Println(buff.String())
+	if verbose {
+		log.Printf("body %s \n",string(buffBytes))
+	}
+	// Send to other servers for view request
+	for i := 0; i < len(target); i++ {
+		// parse the url
+		url, _ := url.Parse(target[i])
+		url.Path = "/view"
+		if verbose {
+			log.Printf("url: %s \n", target[i])
+		}
+		newReqBody := ioutil.NopCloser(bytes.NewBuffer(buffBytes))
+		// reusing requests is unreliable, so copy to new request
+
+		/*
+		res, err := http.DefaultClient.Do(newReq)
+		*/
+		res, err := http.Post(url.String(), "application/json", newReqBody)
+		// Todo: Question: is it alright if we make the request like this? Or do we have to copy over the headers and stuff like before?
+		// Also is Post preferable? or is there not much of a difference?
+		if err != nil {
+			log.Printf("Error when sending request", err)
+		} else {
+			defer res.Body.Close()
+			var responseObj ResponseObj
+			/*
+			fmt.Println("HTTP Response Status:", res.StatusCode, http.StatusText(res.StatusCode))
+			//fmt.Println(res.Body)
+			buf := new(bytes.Buffer)
+			buf.ReadFrom(res.Body)
+			newStr:=buf.String()
+			fmt.Println(newStr)
+		
+			if verbose {
+				for k, v := range res.Header {
+					log.Printf("head: %s, val: %s \n", k, v)
+				}
+			}
+			*/
+			body, bodyErr := ioutil.ReadAll(res.Body)
+			if bodyErr != nil {
+				// Do we want to stop the program here 
+				log.Fatal(bodyErr)
```

Comment:
A bad response shouldn't stop the whole program. We should do either send an error message to the client, throw away the message without telling client, or retry request once more. I think that if there is more than one view requests being sent in a mult-region case and one request gets an invalid response, we just throw it away without telling client.  If all view requests fail, then we should send an error message back to the client.